test/websocket.test.js

```diff
@@ -3857,6 +3857,31 @@ describe('WebSocket', () => {
         agent
       });
     });
+
+    it('supports late headers using finishRequest', (done) => {
+      const wss = new WebSocket.Server({ port: 0 }, () => {
+        const ws = new WebSocket(`ws://localhost:${wss.address().port}`, {
+          finishRequest(req, ws_arg) {
+            setTimeout(() => {
+              assert.strictEqual(ws_arg, ws);
+              assert.strictEqual(req, ws._req);
+              req.setHeader('Cookie', 'foo=bar');
+              req.end();
+            }, 150);
+          }
```

Comment:
```suggestion
          finishRequest(request, websocket) {
            setImmediate(() => {
              assert.strictEqual(request, ws._req);
              assert.strictEqual(websocket, ws);
              req.setHeader('Cookie', 'foo=bar');
              req.end();
            });
          }
```

Reply:
I used setTimeout to verify that it works "later" when everything that can progress without `req.end()` has had a chance to do so, setImmediate will not be sufficient  for that.

Reply:
It doesn't matter we can also do synchronously. We only have to test that a header added here is received. Headers are buffered and not sent (unless `request.flushHeaders()` is called) until `request.end()` is called.

Reply:
I know that delaying it a bit *shouldn't* matter, but I'd prefer the test to confirm that this is genuinely the case. I could alternatively make the test reflect actual expected use:
```js
    it('honors the `finishRequest` option', (done) => {
      const server = https.createServer({
        cert: fs.readFileSync('test/fixtures/certificate.pem'),
        key: fs.readFileSync('test/fixtures/key.pem')
      });
      const wss = new WebSocket.Server({ server });

      wss.on('connection', (ws, req) => {
        assert.strictEqual(req.headers.cookie, 'foo=bar');
        server.close(done);
      });

      server.listen(0, () => {
        const ws = new WebSocket(`wss://localhost:${wss.address().port}`, {
          rejectUnauthorized: false,
          finishRequest(req, ws) {
            assert.strictEqual(req, ws._req);
            req.on('socket', socket => {
              socket.on('secureConnect', () => {
                req.setHeader('Cookie', 'foo=bar');
                req.end();
              });
            });
          }
        });

        ws.on('open', () => {
          ws.close();
        });
      });
    });
```

Also, all other tests abbreviate request to `req` and websocket to `ws`, it would be inconsistent to spell them out here.

Reply:
It would certainly be better than waiting arbitrarily for 150 ms. I don't think the test needs to use TLS, it can use the `'connect'` event of the socket.

> Also, all other tests abbreviate request to `req` and websocket to `ws`, it would be inconsistent to spell them out here.

My suggestion was because I don't like `ws_arg` (snake case) or `wsArg`. `ws` can't be used, so for consistency with the verbose `websocket` I also used the verbose `request`.

Reply:
How about
```js
    it('honors the `finishRequest` option', (done) => {
      const wss = new WebSocket.Server({ port: 0 }, () => {
        const ws = new WebSocket(`ws://localhost:${wss.address().port}`, {
          finishRequest(req, ws) {
            assert.strictEqual(req, ws._req);
            req.on('socket', socket => {
              socket.on('connect', () => {
                req.setHeader('Cookie', 'foo=bar');
                req.end();
              });
            });
          }
        });

        ws.on('open', () => {
          ws.close();
        });
      });

      wss.on('connection', (ws, req) => {
        assert.strictEqual(req.headers.cookie, 'foo=bar');
        wss.close(done);
      });
    });
```
this doesn't explicitly test that the inner `ws` equals the outer one but if the `ws` argument isn't the websocket then the `assert.strictEqual(req, ws._req);` test will catch that anyway.

Reply:
I'm not very happy with it because that is an implementation detail. Anyway my bad, I should have fixed the nits myself after merging.

Reply:
See https://github.com/websockets/ws/commit/23acf8cfaff73fadf89c69be669b3baa29b60233. It is similar to your suggestion with a few more assertions to make sure that the `http.ClientRequest` received by the `finishRequest` function is the same created by the `WebSocket` constructor. In this way with`assert.strictEqual(req, ws._req)` we also make sure that the inner `ws` is the same as the outer `ws` as they share the same request.

There are still no checks for reference equality so it is not perfect but I prefer it to using `process.nextTick()`.